File: src/main/java/com/github/wibowo/OperationExecution.java
package com.github.wibowo;

import com.google.common.collect.Lists;
import org.jetbrains.annotations.NotNull;

import java.util.List;

/**
 * A snapshot of operation execution. Having this snapshot allows us to rollback with exact
 * parameter used to perform the execution. E.g. for {@link Operation#Undo} or when we want to rollback
 * due to arithmetic exception.
 */
public final class OperationExecution {

    /**
     * Operation that was attempted
     */
    private final Operation operation;

    /**
     *  Arguments required by the operation
     */
    private final List<RealNumber> arguments;

    /**
     * Result of the operation execution
     */
    private final RealNumber result;

    public OperationExecution(final @NotNull Operation operation,
                              final @NotNull List<RealNumber> arguments) {
        this.operation = operation;
        this.arguments = arguments;
        this.result = operation.evaluate(this.arguments);
    }

    public OperationExecution(final @NotNull Operation operation,
                              final @NotNull RealNumber arguments) {
        this(operation, Lists.newArrayList(arguments));
    }

    public Operation getOperation() {
        return operation;
    }

    public List<RealNumber> getArguments() {
        return arguments;
    }

    public RealNumber getResult() {
        return result;
    }

    @Override
    public String toString() {
        return result.toString();
    }
}

File: src/main/kotlin/com/github/wibowo/OperationExecution.kt
package com.github.wibowo

/**
 * A snapshot of operation execution. Having this snapshot allows us to rollback with exact
 * parameter used to perform the execution. E.g. for [Operation.Undo] or when we want to rollback
 * due to arithmetic exception.
 *
 * @param operation: Operation that was attempted
 * @param arguments: arguments required by the operation
 */
class OperationExecution(
        val operation: Operation,
        val arguments: List<RealNumber>) {

    /**
     * Result of the operation execution
     */
    val result: RealNumber = operation.evaluate(this.arguments)

    constructor(operation: Operation,
                arguments: RealNumber) : this(operation, listOf(arguments))

    override fun toString(): String {
        return result.toString()
    }
}
